### DISCUSSION TITLE
Welcome to the RStudio Conference People Analytics Workshop Discussion Page!

### DISCUSSION THREAD

**Discussion by keithmcnulty** [rstudio-conf-2022/people-analytics-rstats#2] [Announcements]
## 👋 Welcome!
  We’re using Github Discussions as a place to connect with other participants and ask questions to faculty and to fellow participants. We hope that you use this forum to:
  * Ask questions you’re wondering about or need help on
  * View others' questions and provide help/ideas where you can
  * Provide general comments, feedback or share ideas
 
## Asking questions
We have a lot to cover in the workshop and we won't have a huge amount of time for live discussion during the content modules.  We are also covering some big topics in a shortened amount of time, so there will no doubt be details or concepts that you may want more information on or a deeper explanation.  Whenever you have a question, we encourage you to post it here as a Q&A post.  Faculty will check regularly during the workshop and immediately afterwards and will do their best to respond in as helpful a way as possible.  Other participants should also feel welcome to respond to posts made by others.

## How to start a post
* Click on "Discussions" above
* Click on "New Discussions" 
* Choose a category and go ahead and write your post.

## Tagging instructors
If you want your question or post to be seen by a specific instructor, you can tag @keithmcnulty, @liz-romero or @alex-lopilato in your post.

**Comment by 1melomelo** [comment]
for storing ordered categories, does it matter if you store from low to high or high to low?

**Reply by liz-romero** [reply]
You'll want to choose an order that makes intuitive sense for how you will be interpreting model results.

**Comment by 1melomelo** [comment]
when coercing data structure type, does the conversion always happen by turning the first element into the data type of the 2nd element?

**Reply by liz-romero** [reply]
No, R will choose the datatype that works for all of the elements in the vector.

```
mixed_types_1 <- c(6.2, 8L, FALSE, "Keith")
str(mixed_types_1) # chr


mixed_types_2 <- c(6.2, 8L, FALSE)
str(mixed_types_2) # num

mixed_types_3 <- c(8L, 6.2)
str(mixed_types_3) # num

mixed_types_4 <- c(8L)
str(mixed_types_4) # int
```

**Comment by 1melomelo** [comment]
For Exercise 4, I got 2565. Is that correct? This is the function I made: 

# Using the pipe operator, write code to calculate the average amount
# donated by Rural Domestic individuals, rounded to the nearest dollar.
# Remember that == (double equals) is used for precise equality


donations |> 
  dplyr::filter(reside=="Rural Domestic") |> 
  dplyr::pull(total_donations/n_donations) |> 
  mean() |> 
  round(0)

**Comment by 1melomelo** [comment]
For Exercise 5, I created a histogram using this: 
library(ggplot2)
ggplot(donations, aes(x=total_donations)) + geom_histogram()

I got a plot but had a warning "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`"

**Reply by liz-romero** [reply]
This is just a generic warning.  `geom_histogram()` defaults to 30 bins, but you can choose a binwidth that more appropriately fits your data.  One common approach to calculating binwidths is the [Freedman-Diaconis rule](https://en.wikipedia.org/wiki/Freedman%E2%80%93Diaconis_rule).

**Comment by smithjd** [comment]
what do the 2 decimal places mean in df = 339.75 ?

**Reply by liz-romero** [reply]
Hi, I'm not in the room unfortunately, so I'm not sure how to answer this out of context.  If you can provide more information I can try to answer, or @alex-lopilato would you be able to respond?

**Reply by alex-lopilato** [reply]
By default the t.test() function does not assume that the within-group variances are equal across groups (i.e. the variance in men's ages does not equal the variance in female's ages) and applies a statistical correction that results in an approximate degrees of freedom. If you add the argument var.equal = T to the t.test() function [t.test(age_m, age_f, var.equal = TRUE)], then the correction will not be applied and the df will equal 352 (Total Sample Size [# men + # women] - 2).

**Reply by smithjd** [reply]
Thanks!  Makes sense.

**Comment by 1melomelo** [comment]
Where is link/url to the solutions? Thanks

**Reply by jienagu** [reply]
https://github.com/rstudio-conf-2022/people-analytics-rstats/tree/main/docs/materials/module-exercises All markdown files with SOLUTIONS are the corresponding solutions for each exercise.

**Comment by joshorenstein** [comment]
Do you use ordinal regression for ranked data?

**Reply by keithmcnulty** [reply]
No.  Ranked data has a highly specialized set of modeling approaches (eg Plackett Luce models).  There is a good paper on this here which includes R packages https://scholarworks.umass.edu/cgi/viewcontent.cgi?article=1589&context=pare

**Comment by joshorenstein** [comment]
How do you interpret the odds ratios on the coefficients on the last two variables 1|2, 2|3 in the model from exercise 3 in the ordinal regression?

![image](https://user-images.githubusercontent.com/18400642/180852023-7ff1abbd-187f-416c-bb6e-efc46a981599.png)

**Reply by keithmcnulty** [reply]
These do not have a great deal of practical value.   In theory they represent the baseline odds of moving from 1-2 or from 2-3 assuming that we have no information on the input variables.   You usually have to negate the coefficients in order to achieve this interpretation due to the way the mathematics of the model are constructed.

**Comment by 1melomelo** [comment]
For the KM Estimator that models interactions as well, does it account for the number of groups being compared? Like Bonferonni's correction when comparing many groups? Or does that not apply to KM Estimators or Survival Analysis in general?

**Reply by keithmcnulty** [reply]
@alex-lopilato would be best to answer this I think.  I suspect the answer is no.

**Comment by maia-sh** [comment]
With regards to MER models, I've found this quick visualization/tutorial very helpful: http://mfviz.com/hierarchical-models/

**Comment by AEKEYS** [comment]
Great first day -- thanks, crew! (And hope you're feeling better soon Liz... get plenty of rest and don't push it!).

Open question: I'd be curious to the group's thoughts re: evaluating models using traditional statistical techniques (e.g., model fit measures like R2, pseudo-R2, F statistic, AIC, etc.) versus assessing prediction/forecast errors on a holdout set (e.g., type I, type II errors, FP and FN rates, F1 scores, etc.); I'm sure there's a role for each but would be curious to hear how you all think about that and/or what makes you decide to take the time to assess one or the other (or both).

**Reply by keithmcnulty** [reply]
Generally, train-test splits, confusion matrices, AUC, F1 and all those other predictive analytics metrics have a clear purpose which is to assess your model for the purpose of making accurate predictions of future data.  If your clear primary purpose is to build a model in order to feed data into it for future predictions to be used operationally in decision making, this would be the right approach to validating the model.  These situations are less common in people analytics due to the high stakes setting that such models would be deployed in.

If the primary reason for constructing your model is to understand the potential causes of the outcome or explain to stakeholders what factors influence the outcome, then in general predictive metrics are not so useful, as they give very little precise information on individual input variables and their effect on the outcome.   There is also no need to use a test set since you are not intending to use future data in the model, so you can use all your data set to garner as much explanatory information as possible using the kinds of metrics we have presented in this workshop.  

One fairly rare exception to the above is when you want to explain an outcome and you have an extremely large data set with which to do so (eg 100k+ observations?).   The frequentist statistics methods we have presented can be less effective on very large data sets because even tiny difference can be returned as significant in hypothesis tests when you have very large N.  As I said, these situations are unusual but when I have come across them I have employed non-parametric models and some explainable AI techniques like LIME and SHAP analysis.  These can help you understand the relative importance of input variables but are not great at quantifying their effect.  Another option is to try some Bayesian approaches.

Hope that helps!

**Reply by AEKEYS** [reply]
Thanks for your insights, Keith (and crew)! This (and the rest of the session) was very useful.

So, I guess you're saying that I _should not_ be using the model we trained in 08-Grocery_Performance_Evaluation_Project to auto-magically evaluate all my new employees?? ;-)

`your_compensation <- predict(model, newdata = your_stats)`

**Comment by 1melomelo** [comment]
Any tips on how to save a sociogram image file that's 300 dpi resolution? Thanks

**Reply by liz-romero** [reply]
@jienagu - any ideas for this one?

**Reply by jienagu** [reply]
Yes, @1melomelo . Please use this following code here. `res=300` specifies that you need a figure with a resolution of 300 dpi. 
```
tiff("test.tiff", units="in", width=5, height=5, res=300)
# insert plot code
dev.off()
```

**Reply by 1melomelo** [reply]
Thanks!

**Comment by 1melomelo** [comment]
What's the best layout for networks that have approx. 300 nodes?

**Reply by liz-romero** [reply]
There isn't really 1 best option, different layouts end up working better for certain networks.  I'd recommend checking out the [graph layouts section](https://ona-book.org/viz-graphs.html#layouts) of Keith's network book for a brief overview of some common options.

**Comment by joshorenstein** [comment]
Can you point me to some good practice data for graph analysis?

**Reply by liz-romero** [reply]
[Kaggle](https://www.kaggle.com/datasets?search=network) has lots of options, or here are the datasets from [Keith's book](https://github.com/keithmcnulty/ona_book/tree/main/docs/data)

**Reply by keithmcnulty** [reply]
Also Sociopatterns and the Stanford Network Analysis Project (SNAP) have websites with data.

**Comment by joshorenstein** [comment]
I'm working locally and get a warning message for the centrality betweenness() function - Is there a better function to use?

> Warning: Problem with `mutate()` column `BTWN_CENT`.
> i `BTWN_CENT = tidygraph::centrality_betweenness()`.
> i 'nobigint' is deprecated since igraph 1.3 and will be removed in igraph 1.4

**Reply by keithmcnulty** [reply]
@liz-romero is our tidygraph person - she may have seen this before?

**Reply by liz-romero** [reply]
`tidygraph` wraps functions from `igraph`, so in this case, `tidygraph::centrality_betweenness()` is actually calling `igraph::betweenness()`, including setting a parameter called `nobigint`.  I don't think this is anything to worry about, it's just a warning about an upcoming change but wouldn't impact how we're using the function here.  The [igraph documentation](https://igraph.org/r/doc/betweenness.html) mentions this deprecation as well.

**Comment by joshorenstein** [comment]
@keithmcnulty - I was the one trying to filter on one vertice in the Senator/Twitter problem. I had to miss the last session, but were you able to solve this?

`ego(senator_graph,order = 1, nodes=V(senator_graph)[V(senator_graph)$name== "Paul, Rand"])`

**Reply by keithmcnulty** [reply]
Hi Josh, yes, I think we over complicated it - this should work I think:

`ego(senator_graph,order = 1, nodes="Paul, Rand")`

Let me know if it works!

**Reply by keithmcnulty** [reply]
https://ona-book.org has lots of little examples of these sorts of useful functions in the relevant chapters.

**Comment by juliafsuzuki** [comment]
Hello - Is this the right place to post questions? I don't see the "New Discussions" button / icon / link.

**Reply by keithmcnulty** [reply]
Hi Julia - to start a new discussion, click on the 'Discussions' tab at the top of the repo and you should then see a green 'New Discussion' button.  See example below.
<img width="1332" alt="Screenshot 2022-09-12 at 09 02 35" src="https://user-images.githubusercontent.com/21138031/189603061-21613080-59bf-4f95-a31d-52618060efe0.png">

**Comment by juliafsuzuki** [comment]
Hi Keith, 

 

Hope you had a nice weekend. Thank you for your reply. 

 

Your screen print was helpful so I could compare it with mine.  I have a github account, but don’t see the green ‘New Discussion’ button. Below screen print shows what I see from my end.   

According to the post on github community (link <https://github.com/community/community/discussions/29> ), only those with triage+ permission can open new discussion.  

 

I also have another issue with R cloud. Every time I try to run R codes from the workshop, they stall and they never finish running. The workaround I found is to create a new project and start from scratch.

Not sure if this is intended, but I thought I let you know. 😊 

 

Screen Print #1: The ‘New Discussion’ button is missing



 

Screen Print #2: Created 5 new projects so I can run R codes to see results

**Comment by keithmcnulty** [comment]
Hi Julia, 

I have noticed a bug in some RStudio installations recently where if you click to run a code chunk inside an RMarkdown document, it appears to never finish running.  But if you copy and paste the chunk code into the R console it runs fine.  I have reported this issue.  Is this what you are experiencing?  Unfortunately your screen prints have not come through.   Thanks.

**Comment by juliafsuzuki** [comment]
Yes to "you click to run a code chunk inside an RMarkdown document, and it appears to never finish running"


Screen Print #1: The "New Discussion" button is missing on my page
![PeopleAnalytics-DiscussionPage](https://user-images.githubusercontent.com/77695324/189677008-a4aef1e8-5c36-48b7-93ff-abbb60ff6eeb.JPG)

Screen Print #2: Code chunk (even simple one like below) kept running until I manually terminate
![PeopleAnalytics-KeepRunning](https://user-images.githubusercontent.com/77695324/189678646-c43cfcb7-bb87-4c08-a3ce-1617a1cc1cd5.JPG)

Thank you for reporting the bug you identified!

**Comment by keithmcnulty** [comment]
Hi Julia

The issue you are experiencing seems to be a bug which RStudio will need to
fix.  I have reported this to them, and have seen it in other products as
well.  Re: Github Discussions, it does seem to be permission related.  Feel
free to continue to use this thread instead of starting a new discussion.

Thanks!

On Mon, Sep 12, 2022 at 3:16 PM juliafsuzuki ***@***.***>
wrote:

> Yes to "you click to run a code chunk inside an RMarkdown document, and it
> appears to never finish running"
>
> Not sure if I am not able to attach a screen print...
>
> [image: PeopleAnalytics-DiscussionPage]
> <https://user-images.githubusercontent.com/77695324/189677008-a4aef1e8-5c36-48b7-93ff-abbb60ff6eeb.JPG>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio-conf-2022/people-analytics-rstats/discussions/2#discussioncomment-3626546>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFBIU377PXIQHZUDDVX7PMTV543K7ANCNFSM5WPE7ICA>
> .
> You are receiving this because you were mentioned.Message ID:
> <rstudio-conf-2022/people-analytics-rstats/repo-discussions/2/comments/3626546
> @github.com>
>


-- 
-----------------------------------------------
Keith McNulty
http://www.linkedin.com/in/keith-mcnulty

**Comment by juliafsuzuki** [comment]
Hello - Question regarding 'Structural Equation Models with Latent Variables'.

I ran the Structural Equation Model using a survey data that I collected from people, but get the following error message:
"lavaan ERROR: sample covariance matrix is not positive-definite". To fix this issue, I have the following questions.

Q1. What recommendation do you have for handling Likert values such as 'Not applicable' or 'Neutral / Neither agree or disagree'.
I assigned Likert scales in the following manner. Do you see any issue with this assignment?
0: No applicable
1: Strongly disagree
2: Disagree
3: Neutral
4: Agree
5: Strongly agree

Q2. I might be getting the lavaan error message because my sample size is small. To increase the sample size, can I simply copy and past the existing rows? 10 rows -> 20 rows -> 40 rows -> ... -> 2000 rows?!

Thank you

**Reply by keithmcnulty** [reply]
@alex-lopilato would be best to respond to this when he has time.

**Reply by alex-lopilato** [reply]
Hi Julia, the not positive-definite error you're receiving likely means there is a set of variables (could just be 2 variables) that have very high correlations or that there is a set of variables (could be one or more) that have close to no variance. 

For Q1, I would code "Not applicable" as missing data (NA). For "Neutral" I would do what you've done: code it as 3. 

For Q2, it's very possible you might be getting the error because your sample size is too sample, in which case there isn't much of a fix beyond collecting a larger sample. I wouldn't recommend increasing your sample size by copying and pasting existing rows of data. You're not getting any new information by doing that. 

Hope this helps!

**Reply by juliafsuzuki** [reply]
Got it. Thank you @alex-lopilato for your reply! Hope you are doing well.
